### DISCUSSION TITLE
Weird behaviour of `trigger` in `useEffect`

### DISCUSSION THREAD

**Discussion by mic4ael** [react-hook-form/react-hook-form#7258] [General] [answered]
CodeSandbox: https://codesandbox.io/s/condescending-grothendieck-pzpqq?file=/src/App.js

So in the above code sandbox there are two `useEffect`s - one with `async/await` and the other one without it. If I enable the one without `async/await` I experience an issue in which `reset`ting (then triggering validation) the form with correct values causes the errors to be displayed next to the fields (even though the current values are valid). Only if I use `async/await` (the second `useEffect`) can I experience the expected behaviour (no error messages). What is even funnier is that the [7.13.0-next.3](https://github.com/react-hook-form/react-hook-form/releases/tag/v7.13.0-next.3) version fixes that but I still can't wrap my head around why `async/await` makes the issue go away. I would be very grateful if someone could help me understand this.

**Comment by JamesWong1999** [comment, marked as answer]
I think the problem is due to calling `reset` and `trigger` at the same time. 

related example:

```tsx
import React from "react";
import { useForm, useFieldArray, Controller } from "./src";

function App() {
  const {
    register,
    handleSubmit,
    reset,
    formState: { isSubmitSuccessful }
  } = useForm({ defaultValues: { something: "anything" } });

  const onSubmit = (data) => {
    console.log(data);
  };

  React.useEffect(() => {
    if (formState.isSubmitSuccessful) {
      reset({ something: '' });
    }
  }, [formState, submittedData, reset]);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <input {...register("something")} />
      <input type="submit" />
    </form>
  );
}
```

<img width="953" alt="Screen Shot 2021-12-13 at 11 36 48 am" src="https://user-images.githubusercontent.com/44851854/145736301-13c46a27-6fca-4f7a-a20a-18bacac4ec23.png">

**Comment by mic4ael** [comment]
So I found which line actually fixes the issue - https://github.com/react-hook-form/react-hook-form/pull/6201/files#diff-5a85c1643c71f45e050ce584d7c0db2bdf2e56f8a7dae86d74fedf01be2d2cabR1088. I am still not really sure why `await` before `reset & trigger` prevents the issue from happening. I am sure that for some reason formValues has incorrect value if I don't use `await` thus the validation errors.

**Reply by bluebill1049** [reply]
thanks for looking into the codebase, what do you mean by that line fix the issue?

**Reply by mic4ael** [reply]
With that line, I don't need to use `await` to "fix" the issue.

**Reply by bluebill1049** [reply]
would you like to send a PR?

**Reply by mic4ael** [reply]
No, no. So, the gist of this discussion is to understand why having `await` before `reset` and `trigger` made the issue go away. The line I sent fixed the issue altogether but I would simply like to understand why before `7.13.0-next.3` version I would need to `await` to prevent validation errors.

**Reply by bluebill1049** [reply]
I think  with `await` will result in your function into microtask and hence `register` will be invoked ahead of reset, but what @JamesWong1999 suggested is correct, trigger should/always executed after `useEffect` which means all inputs have been registered after the `reset`

**Reply by bluebill1049** [reply]
Also, the beta version is probably troublesome as well because it set value before reset (include native reset) action

**Reply by mic4ael** [reply]
> I think with await will result in your function into microtask

Yup, indeed, that is the thing. I wondered what could interfere with that. I am aware of the whole microtask queue etc. but I wasn't sure if there is something that is executed before `reset` that could break things.

> hence register will be invoked ahead of reset

You mean that in that case `register` will be called before `reset`, right?

**Reply by bluebill1049** [reply]
> You mean that in that case register will be called before reset, right? 

yes

**Reply by mic4ael** [reply]
> trigger should/always executed after useEffect which means all inputs have been registered after the reset

so in a normal case `register` should be called after `trigger` or before it?

**Reply by bluebill1049** [reply]
trigger should **always** called after inputs mounted (registered)

**Reply by mic4ael** [reply]
> Also, the beta version is probably troublesome as well because it set value before reset (include native reset) action

Why can it be problematic?

**Reply by mic4ael** [reply]
> trigger should always called after inputs mounted (registered)

How can I do that? Can I somehow use `useEffect` to make sure `trigger` is triggerred only after `register`?
